syntax = "proto3";

package raft;

service RaftNode {
  // Raft consensus RPCs
  rpc RequestVote(VoteRequest) returns (VoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  rpc GetLeaderInfo(GetLeaderRequest) returns (GetLeaderResponse);
  
  // User authentication
  rpc Signup(SignupRequest) returns (SignupResponse);
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc Logout(LogoutRequest) returns (StatusResponse);
  
  // Channel operations
  rpc CreateChannel(CreateChannelRequest) returns (StatusResponse);
  rpc GetChannels(GetChannelsRequest) returns (ChannelListResponse);
  rpc JoinChannel(JoinChannelRequest) returns (StatusResponse);
  
  // Messaging
  rpc SendMessage(SendMessageRequest) returns (StatusResponse);
  rpc GetMessages(GetMessagesRequest) returns (MessageListResponse);
  rpc SendDirectMessage(DirectMessageRequest) returns (StatusResponse);
  rpc GetDirectMessages(GetDirectMessagesRequest) returns (DirectMessageListResponse);
  
  // User management
  rpc GetOnlineUsers(GetOnlineUsersRequest) returns (UserListResponse);
  rpc ListConversations(ListConversationsRequest) returns (ConversationsResponse);
  
  // NEW: File operations
  rpc UploadFile(FileUploadRequest) returns (FileUploadResponse);
  rpc DownloadFile(FileDownloadRequest) returns (FileDownloadResponse);
  rpc ListFiles(ListFilesRequest) returns (FileListResponse);
  
  // NEW: LLM features
  rpc GetSmartReply(SmartReplyRequest) returns (SmartReplyResponse);
  rpc SummarizeConversation(SummarizeRequest) returns (SummarizeResponse);
  rpc GetLLMAnswer(LLMRequest) returns (LLMResponse);
  
  // NEW: Channel admin operations
  rpc AddUserToChannel(ChannelAdminRequest) returns (StatusResponse);
  rpc RemoveUserFromChannel(ChannelAdminRequest) returns (StatusResponse);
}

// Raft consensus messages
message VoteRequest {
  int32 term = 1;
  int32 candidate_id = 2;
  int32 last_log_index = 3;
  int32 last_log_term = 4;
}

message VoteResponse {
  int32 term = 1;
  bool vote_granted = 2;
}

message LogEntry {
  int32 term = 1;
  string command = 2;
  bytes data = 3;
}

message AppendEntriesRequest {
  int32 term = 1;
  int32 leader_id = 2;
  int32 prev_log_index = 3;
  int32 prev_log_term = 4;
  repeated LogEntry entries = 5;
  int32 leader_commit = 6;
}


message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
}

message GetLeaderRequest {}

message GetLeaderResponse {
  bool is_leader = 1;
  int32 leader_id = 2;
  string leader_address = 3;
  int32 term = 4;
  string state = 5;
}

// User authentication messages
message SignupRequest {
  string username = 1;
  string password = 2;
  string email = 3;
  string display_name = 4;
}

message SignupResponse {
  bool success = 1;
  string message = 2;
  UserInfo user_info = 3;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  bool success = 1;
  string token = 2;
  string message = 3;
  UserInfo user_info = 4;
}

message LogoutRequest {
  string token = 1;
}

message UserInfo {
  string user_id = 1;
  string username = 2;
  bool is_admin = 3;
  string status = 4;
  string display_name = 5;
  string email = 6;
}

// Channel messages
message CreateChannelRequest {
  string token = 1;
  string channel_name = 2;
  string description = 3;
  bool is_private = 4;
}

message GetChannelsRequest {
  string token = 1;
}

message Channel {
  string channel_id = 1;
  string name = 2;
  string description = 3;
  bool is_private = 4;
  int32 member_count = 5;
}

message ChannelListResponse {
  bool success = 1;
  repeated Channel channels = 2;
}

message JoinChannelRequest {
  string token = 1;
  string channel_id = 2;
}

// Messaging
message SendMessageRequest {
  string token = 1;
  string channel_id = 2;
  string content = 3;
}

message GetMessagesRequest {
  string token = 1;
  string channel_id = 2;
  int32 limit = 3;
  int32 offset = 4;
}

message Message {
  string message_id = 1;
  string sender_id = 2;
  string sender_name = 3;
  string channel_id = 4;
  string content = 5;
  int64 timestamp = 6;
}

message MessageListResponse {
  bool success = 1;
  repeated Message messages = 2;
}

message DirectMessageRequest {
  string token = 1;
  string recipient_username = 2;
  string content = 3;
}

message GetDirectMessagesRequest {
  string token = 1;
  string other_username = 2;
  int32 limit = 3;
  int32 offset = 4;
}

message DirectMessage {
  string message_id = 1;
  string sender_id = 2;
  string sender_name = 3;
  string recipient_id = 4;
  string recipient_name = 5;
  string content = 6;
  int64 timestamp = 7;
  bool is_read = 8;
}

message DirectMessageListResponse {
  bool success = 1;
  repeated DirectMessage messages = 2;
}

// User management
message GetOnlineUsersRequest {
  string token = 1;
}

message UserListResponse {
  bool success = 1;
  repeated UserInfo users = 2;
}

message ListConversationsRequest {
  string token = 1;
}

message Conversation {
  string username = 1;
  string display_name = 2;
  int32 unread_count = 3;
}

message ConversationsResponse {
  bool success = 1;
  repeated Conversation conversations = 2;
}

// NEW: File operations
message FileUploadRequest {
  string token = 1;
  string file_name = 2;
  bytes file_data = 3;
  string channel_id = 4;
  string recipient_username = 5;
  string description = 6;
  string mime_type = 7;
}

message FileUploadResponse {
  bool success = 1;
  string message = 2;
  string file_id = 3;
  string file_url = 4;
}

message FileDownloadRequest {
  string token = 1;
  string file_id = 2;
}

message FileDownloadResponse {
  bool success = 1;
  string file_name = 2;
  bytes file_data = 3;
  string mime_type = 4;
}

message ListFilesRequest {
  string token = 1;
  string channel_id = 2;
}

message FileMetadata {
  string file_id = 1;
  string file_name = 2;
  string uploader_name = 3;
  int64 file_size = 4;
  string mime_type = 5;
  string channel_id = 6;
}

message FileListResponse {
  bool success = 1;
  repeated FileMetadata files = 2;
}

// NEW: LLM features
message SmartReplyRequest {
  string token = 1;
  string channel_id = 2;
  int32 recent_message_count = 3;
}

message SmartReplyResponse {
  bool success = 1;
  repeated string suggestions = 2;
}

message SummarizeRequest {
  string token = 1;
  string channel_id = 2;
  int32 message_count = 3;
}

message SummarizeResponse {
  bool success = 1;
  string summary = 2;
  repeated string key_points = 3;
}

message LLMRequest {
  string token = 1;
  string query = 2;
  repeated string context = 3;
}

message LLMResponse {
  bool success = 1;
  string answer = 2;
}

// NEW: Channel admin operations
message ChannelAdminRequest {
  string token = 1;
  string channel_id = 2;
  string target_username = 3;
}

// Generic status response
message StatusResponse {
  bool success = 1;
  string message = 2;
}